File: 23-Graph/java/09-TopologicalSortDfs.java
import java.util.*;

class Solution {
    public int[] topoSort(int V, List<List<Integer>> adj) {
        boolean[] visited = new boolean[V];
        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < V; i++) {
            if (!visited[i]) {
                dfs(i, adj, visited, stack);
            }
        }

        int[] result = new int[V];
        int idx = 0;
        while (!stack.isEmpty()) {
            result[idx++] = stack.pop();
        }

        return result;
    }

    private void dfs(int node, List<List<Integer>> adj, boolean[] visited, Stack<Integer> stack) {
        visited[node] = true;
        for (int neighbor : adj.get(node)) {
            if (!visited[neighbor]) {
                dfs(neighbor, adj, visited, stack);
            }
        }
        stack.push(node);
    }
}

File: 23-Graph/python/09_topological_sort_dfs.py
from typing import List

class Solution:
    def topoSort(self, V: int, adj: List[List[int]]) -> List[int]:
        visited = [False] * V
        stack = []

        def dfs(node: int):
            visited[node] = True
            for neighbor in adj[node]:
                if not visited[neighbor]:
                    dfs(neighbor)
            stack.append(node)

        for i in range(V):
            if not visited[i]:
                dfs(i)

        return stack[::-1]
